feat: add bandersnatch curve

dragan2234 opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

# Pull Request type

Adding bandersnatch parameters https://eprint.iacr.org/2021/1152.pdf



<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

# What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

# Other information

<!-- Any other information that is important to this PR, such as screenshots of how the component looks before and after the change. -->

feltroidprime commented on hydra/definitions.py:
```diff
@@ -233,6 +235,26 @@ def NAF(x):
+        loop_counter=None,
+        line_function_sparsity=None,
+        final_exp_cofactor=None,
+        fp_generator=3, # todo: check if this is true
```
use the function is_generator in this file to check. If not, try 2,3,4,5,6,7, etc... you will find one

feltroidprime commented on src/fustat/definitions.cairo:
```diff
@@ -118,6 +118,52 @@ namespace secp256r1 {
     const MIN_ONE_D3 = 0x0;
 }
 
+namespace bandersnatch {
```
generate this code automatically with the https://github.com/keep-starknet-strange/garaga/blob/cd329cc873b83841b0828080f6e6db95cae87ef8/hydra/definitions.py#L53-L75 to_cairo_zero method

feltroidprime commented on hydra/definitions.py:
```diff
@@ -233,6 +235,26 @@ def NAF(x):
+        irreducible_polys={},
+        nr_a0=None,
+        nr_a1=None,
+        a=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFE1FB22001,  # See https://neuromancer.sk/std/other/Curve25519 for weirstrass form conversion utilities
```
clean the comment related to curve25519

feltroidprime commented on hydra/definitions.py:
```diff
@@ -233,6 +235,26 @@ def NAF(x):
+        nr_a0=None,
+        nr_a1=None,
+        a=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFE1FB22001,  # See https://neuromancer.sk/std/other/Curve25519 for weirstrass form conversion utilities
+        b=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFB870D2E00001,
```
the a and b parameters seems wrong according to https://eprint.iacr.org/2021/1152.pdf section 3.1

dragan2234 replied:
@feltroidprime 

`a` is −3763200000 mod 52435875175126190479447740508185965837690552500527637822603658699934817984513

same is with `b`

feltroidprime replied:
right

dragan2234 commented:
@feltroidprime Thanks for the review, I have addressed the comments! It should be fine now. add_ec test is passing, but locally I still have some issues with the setup on msm:
```
Exception: Error executing ecip-cli: python3: can't open file '/tools/ecip/main.sage.py': [Errno 2] No such file or directory
```
when I try executing `make run` `msm_ecip.cairo`

feltroidprime commented:
> @feltroidprime Thanks for the review, I have addressed the comments! It should be fine now.
> 
> Strange thing is that msm for 1,2,3 points are fine with bandersantch, but 256 points is failing for some reason

Yes it is expected, circuits have been compiled for up to 3 points for the LHS of the equation to verify https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/src/fustat/precompiled_circuits/ec.cairo#L10 

For the RHS, it supports any number of points as it calls the same circuit in a loop : https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/src/fustat/ec_ops.cairo#L233 but for the LHS we would need to modify the outputs / add an extra circuit to call in a loop to allow the LHS computation for arbitrary number of points https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/hydra/precompiled_circuits/ec.py#L243 , depending on the number of coefficients in sum_dlog_div , see number of coefficients in function of n : https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/hydra/precompiled_circuits/all_circuits.py#L352-L357

For now, please just comment the n=256 line in the test so I can merge this, and we can add it later.

dragan2234 commented:
@feltroidprime Thanks for the explanation, just commented out line for testing 256 points
